### PR TITLE
httpbeast: option to set reusePort

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @dom96 
the next PR (https://github.com/dom96/jester/pull/278) will allow using this in jester

### CODE REVIEW

File: src/httpbeast.nim [hunk 1]
```diff
@@ -54,6 +54,10 @@ type
     domain*: Domain
     numThreads: int
     loggers: seq[Logger]
+    reusePort*: Option[bool]
+      # when false, will fail if --threads is passed or if 2
+      # processes bind to same address/port. We use an Option to avoid
+      # breaking code that relies on `Settings(port: port)`
```
Comment (dom96) [inline src/httpbeast.nim:60]: hm, I think we can live with this breakage. `initSettings` should be used anyway.

**Review by dom96 — CHANGES_REQUESTED** [review]
I guess we are relying on the OS complaining about a port being reused without reusePort enabled? Can we add a nicer error message for that case?

File: src/httpbeast.nim (hunk 1 above)
Comment (timotheecour) [reply src/httpbeast.nim:60]: it's in the wild though... in fact, 1st version of this didn't use an Option and it broke some code. Other than that, i agree that initSettings should be used

**Comment by timotheecour** [comment]
> I guess we are relying on the OS complaining about a port being reused without reusePort enabled? Can we add a nicer error message for that case?

yes, I can make a separate PR to add context info (in particular port) in std/net.bindAddr; this can be done independently of this PR

To add more context to this PR, jester.newSettings sets `reusePort = false`, but this flag (until this PR) wasn't honored:

## before PR:

suppose some service already runs in port 5000 (that you may be unaware of, or that might bind right before you launch your new service); then your run:
`nim r tests/example.nim`
it doesn't fail, but if you run `curl http://0.0.0.0:5000`, it may send requests to the 1st running server instead of the one you just created, which is not a sane default.

## after PR:
it now honors `reusePort = false` (which is the default set in jester.newSettings, which is a good default), and `nim r tests/example.nim` will fail with:

```
/Users/timothee/git_clone/nim/jester/tests/example.nim(3) example
/Users/timothee/git_clone/nim/jester/jester.nim(497) serve
/Users/timothee/git_clone/nim/httpbeast/src/httpbeast.nim(488) run
/Users/timothee/git_clone/nim/httpbeast/src/httpbeast.nim(325) eventLoop
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/net.nim(963) bindAddr
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/includes/oserr.nim(94) raiseOSError
Error: unhandled exception: Address already in use [OSError]
```

it's easy to see by runnning `nim r tests/example.nim` in 2 tabs.

## future work
I'd recommend changing the default in httpbeast too (matching the default in jester which is saner):
```
server.setSockOpt(OptReusePort, settings.reusePort.get(true))
=>
server.setSockOpt(OptReusePort, settings.reusePort.get(false))
```

(i can do this in this PR too but it'd be a breaking change so i'd recommend doing it in another PR instead; at least now, users have a choice)

**Comment by timotheecour** [comment]
@dom96 PTAL

raising when numThreads>1 and reusePort = false just causes more complications, so i did something else instead.

File: src/httpbeast.nim [hunk 2]
```diff
@@ -54,6 +54,9 @@ type
     domain*: Domain
     numThreads: int
     loggers: seq[Logger]
+    reusePort: bool
+      ## controlls whether to fail with "Address already in use".
```
Comment (dom96) [inline src/httpbeast.nim:58]:
```suggestion
      ## controls whether to fail with "Address already in use".
```

File: src/httpbeast.nim [hunk 3]
```diff
@@ -54,6 +54,9 @@ type
...
+      ## This is currently ignored if multiple threads are spawned.
```
Comment (dom96) [inline src/httpbeast.nim:59]:
```suggestion
      ## Setting this to false will raise when `threads` are on.
```

File: src/httpbeast.nim [hunk 4]
```diff
@@ -317,7 +321,7 @@ proc eventLoop(params: (OnRequest, Settings)) =
 
   let server = newSocket(settings.domain)
   server.setSockOpt(OptReuseAddr, true)
-  server.setSockOpt(OptReusePort, true)
+  server.setSockOpt(OptReusePort, settings.reusePort)
```
Comment (dom96) [inline src/httpbeast.nim:324]:
```suggestion
  if compileOption("threads") and not settings.reusePort:
    raise newException(HttpBeastDefect, "--threads:on requires reusePort to be enabled in settings")
  server.setSockOpt(OptReusePort, settings.reusePort)
```

File: src/httpbeast.nim [hunk 5]
```diff
@@ -473,8 +477,13 @@ proc run*(onRequest: OnRequest, settings: Settings) =
...
       var threads = newSeq[Thread[(OnRequest, Settings)]](numThreads)
       for i in 0 ..< numThreads:
+        var settings2 = settings
+        if numThreads > 1: settings2.reusePort = true
+          # TODO: in future work, we can honor `reusePort = false` by
+          # attempting to bind to the port, and then on success spawn the threads
+          # with `reusePort = false`.
         createThread[(OnRequest, Settings)](
-          threads[i], eventLoop, (onRequest, settings)
+          threads[i], eventLoop, (onRequest, settings2)
```
Comment (dom96) [inline src/httpbeast.nim:486]:
```suggestion
          threads[i], eventLoop, (onRequest, settings)
```

File: src/httpbeast.nim [hunk 6]
```diff
@@ -473,8 +477,13 @@ proc run*(onRequest: OnRequest, settings: Settings) =
     when compileOption("threads"):
       var threads = newSeq[Thread[(OnRequest, Settings)]](numThreads)
       for i in 0 ..< numThreads:
+        var settings2 = settings
+        if numThreads > 1: settings2.reusePort = true
+          # TODO: in future work, we can honor `reusePort = false` by
+          # attempting to bind to the port, and then on success spawn the threads
+          # with `reusePort = false`.
```
Comment (dom96) [inline src/httpbeast.nim:484]:
```suggestion
```

**Review by dom96 — APPROVED** [review]

File: src/httpbeast.nim [hunk 7]
```diff
@@ -317,7 +321,9 @@ proc eventLoop(params: (OnRequest, Settings)) =
 
   let server = newSocket(settings.domain)
   server.setSockOpt(OptReuseAddr, true)
-  server.setSockOpt(OptReusePort, true)
+  if compileOption("threads") and not settings.reusePort:
+    raise newException(HttpBeastDefect, "--threads:on requires reusePort to be enabled in settings")
```
Comment (dom96) [inline src/httpbeast.nim:325]:
```suggestion
    raise HttpBeastDefect(msg: "--threads:on requires reusePort to be enabled in settings")
```

File: src/httpbeast.nim [hunk 8]
```diff
@@ -63,14 +66,15 @@ const
 proc initSettings*(port: Port = Port(8080),
                    bindAddr: string = "",
                    numThreads: int = 0,
-                   domain = Domain.AF_INET): Settings =
-
+                   domain = Domain.AF_INET,
+                   reusePort = false): Settings =
```
Comment (dom96) [inline src/httpbeast.nim:70]:
```suggestion
                   reusePort = true): Settings =
```

**Comment by dom96** [comment]
Sorry, I prefer for this stuff to be explicit. So I changed things.

**Pull request closed without merging**